lib/matplotlib/backend_tools.py

```diff
@@ -0,0 +1,927 @@
+
+        # process current tools
+        for tool in self.navigation.tools.values():
+            self._add_tool(tool)
+
+    def _tool_trigger_cbk(self, event):
+        if event.tool.toggled:
+            self._cursor = event.tool.cursor
+        else:
+            self._cursor = None
+
+        self._set_cursor_cbk(event.canvasevent)
+
+    # If the tool is toggleable, set the cursor when the tool is triggered
+    def _add_tool(self, tool):
+        if getattr(tool, 'cursor', None) is not None:
+            self.navigation.nav_connect('tool_trigger_%s' % tool.name,
+                                        self._tool_trigger_cbk)
+
+    # If tool is added, process it
+    def _add_tool_cbk(self, event):
+        if event.tool is self:
+            return
+
+        self._add_tool(event.tool)
+
+    def _set_cursor_cbk(self, event):
+        if not event:
+            return
+
+        if not getattr(event, 'inaxes', False) or not self._cursor:
+            if self._last_cursor != self._default_cursor:
+                self.set_cursor(self._default_cursor)
+                self._last_cursor = self._default_cursor
+        else:
+            if self._cursor:
+                cursor = self._cursor
+                if cursor and self._last_cursor != cursor:
+                    self.set_cursor(cursor)
+                    self._last_cursor = cursor
+
+    def set_cursor(self, cursor):
+        """Set the cursor
+
+        This method has to be implemented per backend
+        """
+        pass
+
+
+class ToolCursorPosition(ToolBase):
+    """Send message with the current pointer position
+
+    This tool runs in the background reporting the position of the cursor
+    """
+    def __init__(self, *args, **kwargs):
+        ToolBase.__init__(self, *args, **kwargs)
+        self._idDrag = self.figure.canvas.mpl_connect(
+            'motion_notify_event', self.send_message)
+
+    def send_message(self, event):
+        """Call `matplotlib.backend_bases.NavigationBase.message_event`"""
+        if self.navigation.messagelock.locked():
+            return
+
+        message = ' '
+
+        if event.inaxes and event.inaxes.get_navigate():
+            try:
+                s = event.inaxes.format_coord(event.xdata, event.ydata)
+            except (ValueError, OverflowError):
+                pass
+            else:
+                message = s
+        self.navigation.message_event(message, self)
+
+
+class RubberbandBase(ToolBase):
+    """Draw and remove rubberband"""
+    def trigger(self, sender, event, data):
+        """Call `draw_rubberband` or `remove_rubberband` based on data"""
+        if not self.figure.canvas.widgetlock.available(sender):
+            return
+        if data is not None:
+            self.draw_rubberband(*data)
+        else:
+            self.remove_rubberband()
+
+    def draw_rubberband(self, *data):
+        """Draw rubberband
+
+        This method must get implemented per backend
```

Comment:
Perhaps we should take advantage of ABC's? I haven't worked with them personally, but my understanding is that it can let you know if a dev failed to implement all the required methods.